### DISCUSSION TITLE
How to decode the clear text of the QR Code?

### DISCUSSION THREAD

**Discussion by Tom-EllisEVENTS** [Digitaler-Impfnachweis/certification-apis#151] [Q&A] [answered]
As far as I understand from the description under "examples", the clear text is BASE45 encoded... but after decoding the sample, it is still not human readable text... can you point me to a document that describes how the content of the QR-code is built?

**Comment by thinkberg** [comment, marked as answer]
Its `BASE45(ZLIB(COSE(CBOR)))`

https://ec.europa.eu/health/sites/default/files/ehealth/docs/digital-green-certificates_v1_en.pdf

**Reply by Tom-EllisEVENTS** [reply]
Ah thank you very much... so the BASE45 decoded content (after "HC1:") is ZLIB compressed?

**Reply by thinkberg** [reply]
Ah, yes, correct: 

```python
#! /usr/bin/env python3
import json
import sys
import zlib

import base45
import cbor2
from cose.messages import CoseMessage

# decode Base45 (remove HC1: prefix)
decoded = base45.b45decode(sys.argv[1][4:])
# decompress using zlib
decompressed = zlib.decompress(decoded)
# decode COSE message (no signature verification done)
cose = CoseMessage.decode(decompressed)
# decode the CBOR encoded payload and print as json
print(json.dumps(cbor2.loads(cose.payload), indent=2))
```

**Reply by Tom-EllisEVENTS** [reply]
Great! Now I only have to collect the libs for my c# project. Thanks a lot!  :-)

**Reply by mlenkeit** [reply]
@Tom11880 maybe this helps: https://github.com/ehn-dcc-development/hcert-dotnet

**Reply by Tom-EllisEVENTS** [reply]
> @Tom11880 maybe this helps: https://github.com/ehn-dcc-development/hcert-dotnet

This looks very promising... thank you very much.

**Reply by Tom-EllisEVENTS** [reply]
Hi Maximilian... funktioniert! Vielen Dank!
